src/backend/access/aocs/aocsam.c

```diff
@@ -1199,21 +1202,28 @@ aocs_gettuple(AOCSScanDesc scan, int64 targrow, TupleTableSlot *slot)
 			}
 		}
 
-		if (ds->blockRowCount <= 0)
+		if (scan->segrowsprocessed == 0)
 			; /* haven't read block */
 		else
 		{
 			/* block was read */
 			rowcount = aocs_block_remaining_rows(ds);
 			Assert(rowcount >= 0);
 
-			if (startrow + rowcount - 1 >= targrow)
+			int64 current_start_row = isanchor ? startrow : ds->blockFirstRowNum;
```

Comment:
I think the variable declaration is better to move to the loop start (to the place where `isanchor` is declared), as in the surrounding code all variables are declared at block start.